includes/migrations/class-wc-stripe-subscriptions-legacy-sepa-tokens-migrator.php

```diff
@@ -0,0 +1,200 @@
+	 * @return WC_Subscription The Subscription object for which its token must be updated.
+	 * @throws \Exception Skip the migration if the request is invalid.
+	 */
+	private function get_subscription_to_migrate( $subscription_id ) {
+		if ( ! class_exists( 'WC_Subscriptions' ) ) {
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. The WooCommerce Subscriptions extension is not active.', $subscription_id ) );
+		}
+
+		$subscription = wcs_get_subscription( $subscription_id );
+
+		if ( ! $subscription ) {
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%d. Subscription not found.', $subscription_id ) );
+		}
+
+		$migrated_legacy_token_id = $subscription->get_meta( self::LEGACY_TOKEN_PAYMENT_METHOD_META_KEY, true );
+
+		if ( ! empty( $migrated_legacy_token_id ) ) {
+			throw new \Exception( sprintf( '---- Skipping migration of subscription #%1$d (%2$s). Token has already been updated.', $subscription_id, $migrated_legacy_token_id ) );
+		}
+
+		return $subscription;
+	}
+
+	/**
+	 * Returns an updated token to be used for the subscription, given the source ID.
+	 *
+	 * If no token is found, it will return the default token for the customer.
+	 *
+	 * @param string  $source_id The Source or Payment Method ID associated with the subscription.
+	 * @param integer $user_id   The WordPress User ID to whom the subscription belongs.
+	 * @throws \Exception If no replacement token is found.
+	 * @return WC_Payment_Token
+	 */
+	private function get_updated_sepa_token_by_source_id( string $source_id, int $user_id ): WC_Payment_Token {
+		$default_token           = false;
+		$updated_sepa_gateway_id = WC_Stripe_UPE_Payment_Gateway::ID . '_' . WC_Stripe_UPE_Payment_Method_Sepa::STRIPE_ID;
+
+		// This method creates an updated token behind the scenes if it doesn't exist.
+		$customer_tokens = WC_Payment_Tokens::get_customer_tokens( $user_id, $updated_sepa_gateway_id );
+
+		foreach ( $customer_tokens as $token ) {
+			// Return the token once we find it.
+			if ( $source_id === $token->get_token() ) {
+				return $token;
+			}
+
+			// Let's store the default token in case we don't find the one we're looking for.
+			if ( $token->is_default() ) {
+				$default_token = $token;
+			}
+		}
+
+		// We can't proceed with updating the subscription if we don't have a token to use.
+		if ( ! $default_token ) {
+			throw new \Exception( '---- Skipping migration of subscription. No replacement token was found.' );
+		}
+
+		return $default_token;
+	}
+
+	/**
+	 * Sets the updated payment method for the subscription.
+	 *
+	 * @param WC_Subscription  $subscription Subscription for which the payment method must be updated.
+	 * @param WC_Payment_Token $token        The token to be set as the payment method for the subscription.
+	 */
+	private function set_subscription_updated_payment_method( WC_Subscription $subscription, WC_Payment_Token $token ) {
```

Comment:
Hey @mattallan ! Quick question here. Would this be the right way to update the payment method associated with the subscription?

This is what we do in [WC_Stripe_Subscriptions_Trait::maybe_update_source_on_subscription_order()](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/compat/trait-wc-stripe-subscriptions.php#L427), but wondering if we should use WCS_Payment_Tokens::update_subscription_token() instead, or WC_Order::add_payment_token(). The latter is the one used by WooPayments.

The renewals work fine when updating the payment method like this, but maybe I'm missing something.

Reply:
Good question! What you have here is perfectly fine and IMO a bit better than just using `WCS_Payment_Tokens::update_subscription_token()` which doesn't call `set_payment_method()` to ensure the correct Stripe SEPA gateway ID is stored as the subscription.

Using the `WCS_Payment_Tokens::update_subscription_token()` will still work and has some added benefits like updating the latest renewal order with the updated token and also trigger the `'woocommerce_subscription_token_changed'` but I don't think these are necessary to the migrator, are they?

I did a quick search for 'woocommerce_subscription_token_changed' across Woo and Automattic orgs and can't see any plugins using it so I think it's good to go with what you have here.

Curious what @james-allan thinks!

Reply:
I think `set_payment_method()` is good. It's the one we also use in things like the REST API.